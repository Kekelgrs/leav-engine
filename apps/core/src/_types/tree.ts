// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {ITreeNodePermissionsConf} from './permissions';
import {IRecord} from './record';
import {IKeyValue} from './shared';

export interface ITreeLibrarySettings {
    allowMultiplePositions: boolean;
}

export interface ITree extends ICoreEntity {
    libraries: IKeyValue<ITreeLibrarySettings>;
    behavior?: TreeBehavior;
    system?: boolean;
    permissions_conf?: ITreeNodePermissionsConf;
}

export interface ITreeFilterOptions extends ICoreEntityFilterOptions {
    system?: boolean;
}

export interface ITreeElement {
    id: string;
    library: string;
}

export interface ITreeNode {
    order?: number;
    record?: IRecord;
    parent?: ITreeNode[];
    ancestors?: TreePaths;
    children?: ITreeNode[];
    linkedRecords?: IRecord[];
}

export type TreePaths = ITreeNode[][];

export enum TreeBehavior {
    STANDARD = 'standard',
    FILES = 'files'
}
