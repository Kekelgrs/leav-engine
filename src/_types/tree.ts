import {IRecord} from './record';
import {ISystemTranslation} from './systemTranslation';

export interface ITree {
    id: string;
    libraries: string[];
    behavior?: TreeBehavior;
    label: ISystemTranslation;
    system?: boolean;
}

export interface ITreeFilterOptions {
    id?: string;
    label?: string;
    system?: boolean;
}

export interface ITreeElement {
    id: number;
    library: string;
}

export interface ITreeNode {
    order?: number;
    record?: IRecord;
    parent?: ITreeNode[];
    ancestors?: ITreeNode[];
    children?: ITreeNode[];
    linkedRecords?: IRecord[];
}

export enum TreeBehavior {
    STANDARD = 'standard',
    FILES = 'files'
}
