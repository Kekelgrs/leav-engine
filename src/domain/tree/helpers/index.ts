export {default as treeDataValidation} from './treeDataValidation';
