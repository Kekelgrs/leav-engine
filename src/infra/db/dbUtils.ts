import {aql} from 'arangojs';
import {GeneratedAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {CollectionType} from 'arangojs/lib/cjs/collection';
import {AwilixContainer} from 'awilix';
import {readdirSync} from 'fs';
import {IPluginsRepo} from 'infra/plugins/pluginsRepo';
import * as path from 'path';
import {isArray} from 'util';
import * as winston from 'winston';
import {IAttribute, IAttributeFilterOptions} from '_types/attribute';
import {IForm} from '_types/forms';
import {ILibrary, ILibraryFilterOptions} from '_types/library';
import {IList, IPaginationParams, ISortParams} from '_types/list';
import {IQueryInfos} from '_types/queryInfos';
import {IKeyValue} from '_types/shared';
import {ITree, ITreeFilterOptions} from '_types/tree';
import {IDbValueVersion, IValueVersion} from '_types/value';
import {collectionTypes, IDbService, IExecuteWithCount} from './dbService';
import runMigrationFiles from './helpers/runMigrationFiles';

export const MIGRATIONS_COLLECTION_NAME = 'core_db_migrations';

export interface IFindCoreEntityParams {
    collectionName: string;
    filters?: ITreeFilterOptions | ILibraryFilterOptions | IAttributeFilterOptions;
    strictFilters?: boolean;
    withCount?: boolean;
    pagination?: IPaginationParams;
    sort?: ISortParams;
    customFilterConditions?: IKeyValue<
        (filterKey: string, filterVal: string | boolean | string[], strictFilters: boolean) => GeneratedAqlQuery
    >;
    ctx: IQueryInfos;
}

export interface IDbUtils {
    migrate?(depsManager: AwilixContainer): Promise<void>;
    cleanup?(record: {}): any;
    convertToDoc?(obj: {}): any;
    isCollectionExists?(name: string): Promise<boolean>;
    findCoreEntity?<T extends ITree | ILibrary | IAttribute | IForm>(params: IFindCoreEntityParams): Promise<IList<T>>;
    convertValueVersionToDb?(version: IValueVersion): IDbValueVersion;
    convertValueVersionFromDb?(version: IDbValueVersion): IValueVersion;
    clearDatabase(): Promise<void>;
}

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.utils.logger'?: winston.Winston;
    'core.infra.plugins'?: IPluginsRepo;
    config?: any;
}

export default function({
    'core.infra.db.dbService': dbService = null,
    'core.utils.logger': logger = null,
    'core.infra.plugins': pluginsRepo = null,
    config = null
}: IDeps = {}): IDbUtils {
    /**
     * Create the collections used to managed db migrations
     * This can't be in a migration file because we do have to initialize it somewhere
     *
     */
    async function _initMigrationsCollection(): Promise<void> {
        const collections = await dbService.db.listCollections();

        const colExists = collections.reduce((exists, c) => exists || c.name === MIGRATIONS_COLLECTION_NAME, false);
        if (!colExists) {
            const collection = dbService.db.collection(MIGRATIONS_COLLECTION_NAME);
            await collection.create();
        }
    }

    /**
     * Return the filter's conditions based on key and val supplied.
     *
     * @param filterKey
     * @param filterVal
     * @param bindVars
     * @param index
     * @param strictFilters
     */
    function _getFilterCondition(
        filterKey: string,
        filterVal: string | boolean | string[],
        strictFilters: boolean
    ): GeneratedAqlQuery {
        const queryParts = [];

        // If value is an array (types or formats for example),
        // we call this function recursively on array and join filters with an OR
        if (Array.isArray(filterVal)) {
            if (filterVal.length) {
                const valParts = filterVal.map(val => _getFilterCondition(filterKey, val, strictFilters));
                queryParts.push(aql.join(valParts, ' OR '));
            }
        } else {
            if (filterKey === 'label') {
                // Search for label in any language
                const valParts = config.lang.available.map(l => aql`LIKE(el.label.${l}, ${filterVal}, true)`);
                queryParts.push(aql.join(valParts, ' OR '));
            } else {
                // Filter with a "like" on ID or exact value in other fields
                queryParts.push(
                    filterKey === '_key' && !strictFilters
                        ? aql`LIKE(el.${filterKey}, ${filterVal}, true)`
                        : aql`el.${filterKey} == ${filterVal}`
                );
            }
        }

        return aql.join(queryParts);
    }

    const ret = {
        /**
         * Run database migrations.
         * It takes all files present in migrations folder and run it if it's never been executed before
         *
         * @param depsManager
         */
        async migrate(depsManager: AwilixContainer): Promise<void> {
            await _initMigrationsCollection();
            const ctx: IQueryInfos = {
                userId: '0',
                queryId: 'run-migrations'
            };
            // Load already ran migrations
            const executedMigrations = await dbService.execute({
                query: `
                    FOR m IN core_db_migrations
                    RETURN m.file
                `,
                ctx
            });

            const _runMigrationFiles = (files, folder, prefix = null) =>
                runMigrationFiles({
                    files,
                    executedMigrations,
                    migrationsDir: folder,
                    prefix,
                    deps: {depsManager, dbService, logger},
                    ctx
                });

            /*** Core migrations ***/
            // Load migrations files
            const migrationsDir = path.resolve(__dirname, 'migrations');
            const migrationFiles = readdirSync(migrationsDir).filter(file => file.indexOf('.map') === -1);

            await _runMigrationFiles(migrationFiles, migrationsDir);

            /*** Plugins migrations ***/
            const plugins = pluginsRepo.getRegisteredPlugins();
            for (const plugin of plugins) {
                const pluginMigrationFolder = path.resolve(plugin.path + '/infra/db/migrations');
                const pluginMigrationFiles = readdirSync(pluginMigrationFolder).filter(
                    file => file.indexOf('.map') === -1
                );

                await _runMigrationFiles(pluginMigrationFiles, pluginMigrationFolder, plugin.infos.name);
            }
        },

        /**
         * Cleanup every system keys from an object coming from database.
         * _key is kept under 'id'
         *
         * @param obj
         * @return any   Cleaned up object
         */
        cleanup(obj: any): any {
            if (obj === null) {
                return null;
            }

            return Object.keys(obj).reduce((newObj: any, key) => {
                if (key === '_key') {
                    newObj.id = obj[key];
                } else if (key[0] !== '_') {
                    newObj[key] = obj[key];
                }

                return newObj;
            }, {});
        },

        /**
         * Convert an object to an object looking like a DB document
         * id is replaced by _key
         *
         * @param obj
         * @return any   DB document compatible object
         */
        convertToDoc(obj: {}): any {
            const newObj: any = {...obj};

            if (typeof newObj.id !== 'undefined') {
                newObj._key = newObj.id;
            }
            delete newObj.id;

            return newObj;
        },
        /**
         * Search core entities (libraries, attributes, trees)
         *
         * @param collectionName
         * @param filters
         * @param strictFilters
         */
        async findCoreEntity<T extends ITree | ILibrary | IAttribute>(
            params: IFindCoreEntityParams
        ): Promise<IList<T>> {
            const {
                collectionName = null,
                filters = null,
                strictFilters = false,
                withCount = false,
                pagination = null,
                sort = null,
                customFilterConditions = {},
                ctx = {}
            } = params;

            const collec = dbService.db.collection(collectionName);
            const queryParts = [aql`FOR el IN ${collec}`];

            if (filters !== null) {
                const dbFilters = ret.convertToDoc(filters);
                const filtersKeys = Object.keys(dbFilters);

                for (const filterKey of filtersKeys) {
                    const filterVal = dbFilters[filterKey];

                    // Caller can define some custom functions to generate filter condition (like looking on edges to
                    // filter on libraries linked to an attribute). So, if a custom function is define for this filter,
                    // we use it, otherwise we use the standard filters
                    const filterCondsFunc =
                        typeof customFilterConditions[filterKey] !== 'undefined'
                            ? customFilterConditions[filterKey]
                            : _getFilterCondition;
                    const conds = filterCondsFunc(filterKey, filterVal, strictFilters);

                    if (conds.query) {
                        queryParts.push(aql`FILTER`, conds);
                    }
                }
            }

            if (!!sort) {
                const field = sort.field === 'id' ? '_key' : sort.field;
                queryParts.push(aql`SORT el.${field} ${sort.order}`);
            }

            if (!!pagination) {
                queryParts.push(aql`LIMIT ${pagination.offset || 0}, ${pagination.limit}`);
            }

            queryParts.push(aql`RETURN el`);

            const query = aql.join(queryParts);
            const res = await dbService.execute<IExecuteWithCount | any[]>({query, withTotalCount: withCount, ctx});

            const results = !isArray(res) ? res.results : res;

            return {
                totalCount: withCount ? (res as IExecuteWithCount).totalCount : null,
                list: results.map(ret.cleanup)
            };
        },
        convertValueVersionToDb(version: IValueVersion): IDbValueVersion {
            return Object.keys(version).reduce((allVers, treeName) => {
                const {library, id} = version[treeName];

                allVers[treeName] = `${library}/${id}`;

                return allVers;
            }, {});
        },
        convertValueVersionFromDb(version: IDbValueVersion): IValueVersion {
            return version
                ? Object.keys(version).reduce((allVers, treeName) => {
                      const [library, id] = version[treeName].split('/');

                      allVers[treeName] = {
                          library,
                          id
                      };

                      return allVers;
                  }, {})
                : null;
        },
        async clearDatabase(): Promise<void> {
            // Drop all collections
            const cols = await dbService.db.listCollections();
            for (const col of cols) {
                const colType =
                    col.type === CollectionType.DOCUMENT_COLLECTION ? collectionTypes.DOCUMENT : collectionTypes.EDGE;
                await dbService.dropCollection(col.name, colType);
            }
        }
    };

    return ret;
}
