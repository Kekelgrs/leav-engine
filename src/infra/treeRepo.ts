import {ITree, ITreeElement, ITreeFilterOptions, ITreeNode} from '_types/tree';
import {aql} from 'arangojs';
import {IDbService, collectionTypes} from './db/dbService';
import {IDbUtils} from './db/dbUtils';
import {IRecord} from '_types/record';
import {VALUES_LINKS_COLLECTION} from './recordRepo';

export interface ITreeRepo {
    createTree(treeData: ITree): Promise<ITree>;
    updateTree(treeData: ITree): Promise<ITree>;
    getTrees(filters?: ITreeFilterOptions): Promise<ITree[]>;
    deleteTree(id: string): Promise<ITree>;

    /**
     * Add an element to the tree
     *
     * @param element
     * @param parent Parent must be a record or null to add element to root
     */
    addElement(treeId: string, element: ITreeElement, parent: ITreeElement | null): Promise<ITreeElement>;

    /**
     * Move an element in the tree
     *
     * @param element
     * @param parentFrom A record or null to move from root
     * @param parentTo A record or null to move to root
     */
    moveElement(treeId: string, element: ITreeElement, parentTo: ITreeElement | null): Promise<ITreeElement>;

    /**
     * Delete an element from the tree
     *
     * @param element
     * @param parent A record or null to delete from root
     */
    deleteElement(treeId: string, element: ITreeElement, deleteChildren: boolean | null): Promise<ITreeElement>;

    /**
     * Return whether an element is present in given tree
     *
     * @param treeId
     * @param element
     */
    isElementPresent(treeId: string, element: ITreeElement): Promise<boolean>;

    /**
     * Return the whole tree in the form:
     * [
     *      {
     *          record: {...},
     *          children: [
     *              {
     *                  record: ...
     *                  children: ...
     *              }
     *          ]
     *      },
     *      { ... }
     * ]
     *
     * @param treeId
     * @param startingNode  Return the tree starting from this node. If not specified, start from root
     */
    getTreeContent(treeId: string, startingNode?: ITreeElement): Promise<ITreeNode[]>;

    /**
     * Return all first level children of an element
     *
     * @param treeId
     * @param element
     */
    getElementChildren(treeId: string, element: ITreeElement): Promise<ITreeNode[]>;

    /**
     * Return all ancestors of an element, including element itself, but excluding tree root
     *
     * @param treeId
     * @param element
     */
    getElementAncestors(treeId: string, element: ITreeElement): Promise<ITreeNode[]>;

    getLinkedRecords(treeId: string, attribute: string, element: ITreeElement): Promise<IRecord[]>;
}

const TREES_COLLECTION_NAME = 'core_trees';
const EDGE_COLLEC_PREFIX = 'core_edge_tree_';
const MAX_TREE_DEPTH = 1000;

export default function(dbService: IDbService, dbUtils: IDbUtils): ITreeRepo {
    function _getRootId(treeId: string): string {
        return `${TREES_COLLECTION_NAME}/${treeId}`;
    }

    function _getTreeEdgeCollectionName(treeId: string): string {
        return EDGE_COLLEC_PREFIX + treeId;
    }

    return {
        async createTree(treeData: ITree): Promise<ITree> {
            const defaultParams = {_key: '', system: false, label: {fr: '', en: ''}};
            const collec = dbService.db.collection(TREES_COLLECTION_NAME);

            const docToInsert = {...defaultParams, ...dbUtils.convertToDoc(treeData)};

            const treeRes = await dbService.execute(aql`INSERT ${docToInsert} IN ${collec} RETURN NEW`);

            const edgeCollec = await dbService.createCollection(EDGE_COLLEC_PREFIX + treeData.id, collectionTypes.EDGE);

            return dbUtils.cleanup(treeRes.pop());
        },
        async updateTree(treeData: ITree): Promise<ITree> {
            const collec = dbService.db.collection(TREES_COLLECTION_NAME);

            const docToSave = dbUtils.convertToDoc(treeData);

            const treeRes = await dbService.execute(aql`UPDATE ${docToSave} IN ${collec} RETURN NEW`);

            return dbUtils.cleanup(treeRes.pop());
        },
        async getTrees(filters?: ITreeFilterOptions): Promise<ITree[]> {
            let query = `FOR l IN ${TREES_COLLECTION_NAME}`;
            const bindVars = {};

            if (typeof filters !== 'undefined') {
                const dbFilters = dbUtils.convertToDoc(filters);
                const filtersKeys = Object.keys(dbFilters);
                for (let i = 0; i < filtersKeys.length; i++) {
                    query += ` FILTER l.@filterKey${i} == @filterValue${i}`;
                    bindVars[`filterKey${i}`] = filtersKeys[i];
                    bindVars[`filterValue${i}`] = `${dbFilters[filtersKeys[i]]}`;
                }
            }

            query += ` RETURN l`;

            const res = await dbService.execute({query, bindVars});

            return res.map(dbUtils.cleanup);
        },
        async deleteTree(id: string): Promise<ITree> {
            const collec = dbService.db.collection(TREES_COLLECTION_NAME);
            const res = await dbService.execute(aql`REMOVE ${{_key: id}} IN ${collec} RETURN OLD`);

            const delCollec = await dbService.dropCollection(EDGE_COLLEC_PREFIX + id, collectionTypes.EDGE);

            // Return deleted library
            return dbUtils.cleanup(res.pop());
        },
        async addElement(treeId: string, element: ITreeElement, parent: ITreeElement | null): Promise<ITreeElement> {
            const destination = parent !== null ? `${parent.library}/${parent.id}` : _getRootId(treeId);
            const elemId = `${element.library}/${element.id}`;

            const collec = dbService.db.edgeCollection(_getTreeEdgeCollectionName(treeId));

            const res = await dbService.execute(
                aql`INSERT {_from: ${destination}, _to: ${elemId}} IN ${collec} RETURN NEW`
            );

            return element;
        },
        async moveElement(treeId: string, element: ITreeElement, parentTo: ITreeElement | null): Promise<ITreeElement> {
            const destination = parentTo !== null ? `${parentTo.library}/${parentTo.id}` : _getRootId(treeId);
            const elemId = `${element.library}/${element.id}`;

            const collec = dbService.db.edgeCollection(_getTreeEdgeCollectionName(treeId));

            const res = await dbService.execute(
                aql`
                    FOR e IN ${collec}
                        FILTER e._to == ${elemId}
                        UPDATE e WITH {_from: ${destination}, _to: ${elemId}}
                        IN ${collec}
                        RETURN NEW
                `
            );

            return element;
        },
        async deleteElement(
            treeId: string,
            element: ITreeElement,
            deleteChildren: boolean | null = true
        ): Promise<ITreeElement> {
            const collec = dbService.db.edgeCollection(_getTreeEdgeCollectionName(treeId));
            const elemId = `${element.library}/${element.id}`;

            if (deleteChildren) {
                // Remove all element's children
                const resRemChildren = await dbService.execute(aql`
                    LET edges = (
                        FOR v, e IN 0..${MAX_TREE_DEPTH} OUTBOUND ${elemId}
                        ${collec}
                        RETURN e
                    )
                    FOR ed IN edges
                        FILTER ed != null
                        REMOVE ed IN ${collec}
                        RETURN OLD
                `);
            } else {
                const parentId = await dbService.execute(aql`
                    FOR v IN 1 INBOUND ${elemId}
                    ${collec}
                    RETURN v._id
                `);

                const parent = {
                    library: parentId[0].split('/')[0],
                    id: parentId[0].split('/')[1]
                };

                const children = await dbService.execute(aql`
                    FOR v IN 1 OUTBOUND ${elemId}
                    ${collec}
                    RETURN v
                `);

                // Move children to element's parent
                await Promise.all(
                    children.map(child => {
                        const childLib = child._id.split('/')[0];
                        const childId = child._id.split('/')[1];

                        return this.moveElement(treeId, {id: childId, library: childLib}, parent);
                    })
                );
            }

            // Remove element from its parent
            const resRemElem = await dbService.execute(aql`
                FOR e IN ${collec}
                    FILTER e._to == ${elemId}
                    REMOVE e IN ${collec}
                    RETURN OLD
            `);

            return element;
        },
        async isElementPresent(treeId: string, element: ITreeElement): Promise<boolean> {
            const collec = dbService.db.edgeCollection(_getTreeEdgeCollectionName(treeId));
            const elemId = `${element.library}/${element.id}`;

            const res = await dbService.execute(aql`
                FOR e IN ${collec}
                    FILTER e._to == ${elemId}
                    RETURN e
            `);

            return !!res.length;
        },
        async getTreeContent(treeId: string, startingNode?: ITreeElement): Promise<ITreeNode[]> {
            const rootId = _getRootId(treeId);

            const treeContent: ITreeNode[] = [];

            const collec = dbService.db.edgeCollection(_getTreeEdgeCollectionName(treeId));

            const nodeFrom = !!startingNode ? `${startingNode.library}/${startingNode.id}` : rootId;

            /**
             * This query return a list of all records present in the tree with their path IDs
             * from the root. Query is made depth-first
             */
            const data = await dbService.execute(aql`
                FOR v, e, p IN 1..${MAX_TREE_DEPTH} OUTBOUND ${nodeFrom}
                ${collec}
                LET path = (FOR pv IN p.vertices RETURN pv._id)
                RETURN MERGE(v, {path})
            `);

            /**
             * Process query result to transform it to a proper tree structure
             * For each record of the tree, we run through its path to find out where is its parent
             * in the tree.
             * Then we can add it to its parent children.
             */
            for (const elem of data) {
                // We don't want the root in the tree, skip it
                if (elem._id === rootId) {
                    continue;
                }

                // Last path part is the element itself, we don't need it
                elem.path.pop();

                /** Determine where is the parent in the tree */
                let parentInTree: any = treeContent;
                for (const pathPart of elem.path) {
                    // Root's first level children will be directly added to the tree
                    if (pathPart === nodeFrom) {
                        parentInTree = treeContent;
                    } else {
                        // If previous parent was the tree root, there's no 'children'
                        const container = parentInTree === treeContent ? parentInTree : parentInTree.children;

                        // Look for the path to follow among all nodes
                        parentInTree = container.find(el => {
                            const pathLib = pathPart.split('/')[0];
                            const pathId = pathPart.split('/')[1];
                            return el.record.library === pathLib && el.record.id === pathId;
                        });
                    }
                }

                /** Add element to its parent */
                delete elem.path;
                elem.library = elem._id.split('/')[0];

                // If destination is the tree root, there's no 'children'
                const destination = parentInTree === treeContent ? parentInTree : parentInTree.children;

                destination.push({record: dbUtils.cleanup(elem), children: []});
            }

            return treeContent;
        },
        async getElementChildren(treeId: string, element: ITreeElement): Promise<ITreeNode[]> {
            const treeEdgeCollec = dbService.db.edgeCollection(_getTreeEdgeCollectionName(treeId));
            const query = aql`
                FOR v
                    IN 1 OUTBOUND ${element.library + '/' + element.id}
                    ${treeEdgeCollec}
                    RETURN v
            `;

            const res = await dbService.execute(query);
            return res.map(elem => {
                elem.library = elem._id.split('/')[0];
                return {record: dbUtils.cleanup(elem)};
            });
        },
        async getElementAncestors(treeId: string, element: ITreeElement): Promise<ITreeNode[]> {
            const treeEdgeCollec = dbService.db.edgeCollection(_getTreeEdgeCollectionName(treeId));

            const query = aql`
                FOR v,e,p
                    IN 0..${MAX_TREE_DEPTH} INBOUND ${element.library + '/' + element.id}
                    ${treeEdgeCollec}
                    SORT COUNT(p.edges) DESC
                    FILTER v._id != ${_getRootId(treeId)}
                    RETURN v
            `;

            const res = await dbService.execute(query);

            return res.map(elem => {
                elem.library = elem._id.split('/')[0];

                return {record: dbUtils.cleanup(elem)};
            });
        },
        async getLinkedRecords(treeId: string, attribute: string, element: ITreeElement): Promise<IRecord[]> {
            const edgeCollec = dbService.db.edgeCollection(VALUES_LINKS_COLLECTION);

            const query = aql`
                FOR v,e,p
                    IN 1 INBOUND ${element.library + '/' + element.id}
                    ${edgeCollec}
                    FILTER e.attribute == ${attribute}
                    RETURN v
            `;

            const res = await dbService.execute(query);

            return res.map(elem => {
                elem.library = elem._id.split('/')[0];

                return dbUtils.cleanup(elem);
            });
        }
    };
}
