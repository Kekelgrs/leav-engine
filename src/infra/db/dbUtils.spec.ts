import {Database} from 'arangojs';
import {TREES_COLLECTION_NAME} from '../../infra/tree/treeRepo';
import dbUtils from './dbUtils';

describe('dbUtils', () => {
    describe('cleanupSystemKeys', () => {
        test('Should remove all system keys', () => {
            const testDbUtils = dbUtils(null, null);

            const testObj = {
                _key: 'testKey',
                _id: 'testId',
                _rev: 'testRev',
                _randomSystemKey: 'test',
                normalKey: 'shouldBeKept'
            };

            const res = testDbUtils.cleanup(testObj);

            expect(res).toMatchObject({id: 'testKey', normalKey: 'shouldBeKept'});
        });

        test('Should return null if param is null', () => {
            const testDbUtils = dbUtils(null, null);

            const res = testDbUtils.cleanup(null);

            expect(res).toBeNull();
        });
    });

    describe('convertToDoc', () => {
        test('Should add needed system keys', () => {
            const testDbUtils = dbUtils(null, null);

            const testObj = {
                id: 'testId',
                normalKey: 'shouldBeKept'
            };

            const res = testDbUtils.convertToDoc(testObj);

            expect(res).toMatchObject({_key: 'testId', normalKey: 'shouldBeKept'});
        });
    });

    describe('findCoreEntity', () => {
        let mockDbServ;
        let testDbUtils;
        beforeEach(() => {
            mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([
                    {
                        _key: 'categories',
                        _id: 'core_trees/categories',
                        _rev: '_Wm_Qdtu--_',
                        label: {
                            fr: 'Arbre des catégories'
                        },
                        libraries: ['categories'],
                        system: false
                    }
                ])
            };
            testDbUtils = dbUtils(mockDbServ, null, {lang: {available: ['fr', 'en']}});
            testDbUtils.cleanup = jest.fn().mockReturnValue({
                id: 'categories',
                system: false,
                label: {
                    fr: 'Arbre des catégories'
                }
            });
            testDbUtils.convertToDoc = jest.fn().mockReturnValue({
                _key: 'categories',
                system: false,
                label: 'Arbre des catégories'
            });
        });

        test('Find core entity without filters', async () => {
            const res = await testDbUtils.findCoreEntity(TREES_COLLECTION_NAME);

            expect(res).toHaveLength(1);

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatch(/FOR el IN core_trees/);
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].bindVars).toMatchSnapshot();

            expect(res[0]).toMatchObject({
                id: 'categories',
                system: false,
                label: {
                    fr: 'Arbre des catégories'
                }
            });
        });

        test('Filter with a LIKE on ID', async function() {
            const res = await testDbUtils.findCoreEntity(TREES_COLLECTION_NAME, {id: 'test'});

            expect(mockDbServ.execute.mock.calls[0][0].query).toMatch(/(FILTER LIKE){1}/);
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].bindVars).toMatchSnapshot();
        });

        test('Should filter label on any language', async function() {
            const res = await testDbUtils.findCoreEntity(TREES_COLLECTION_NAME, {label: 'test'});

            expect(mockDbServ.execute.mock.calls[0][0].query).toMatch(/(LIKE(.*)label\.fr(.*)OR LIKE(.*)label\.en)/);
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].bindVars).toMatchSnapshot();
        });

        test('Should return an empty array if no results', async function() {
            mockDbServ = {db: null, execute: global.__mockPromise([])};
            testDbUtils = dbUtils(mockDbServ, null, {lang: {available: ['fr', 'en']}});
            testDbUtils.cleanup = jest.fn();
            testDbUtils.convertToDoc = jest.fn();

            mockDbServ = {db: null, execute: global.__mockPromise([])};
            const res = await testDbUtils.findCoreEntity(TREES_COLLECTION_NAME);

            expect(res).toBeInstanceOf(Array);
            expect(res.length).toBe(0);
        });
    });
});
