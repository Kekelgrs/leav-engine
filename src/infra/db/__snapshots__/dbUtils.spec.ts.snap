// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dbUtils findCoreEntity Filter with a LIKE on ID 1`] = `"FOR el IN core_trees FILTER LIKE(el.@filterKey0, @filterValue0, true) FILTER el.@filterKey1 == @filterValue1 FILTER LIKE(el.label.fr, @filterValue2, true) OR LIKE(el.label.en, @filterValue2, true) RETURN el"`;

exports[`dbUtils findCoreEntity Filter with a LIKE on ID 2`] = `
Object {
  "filterKey0": "_key",
  "filterKey1": "system",
  "filterValue0": "categories",
  "filterValue1": false,
  "filterValue2": "Arbre des catégories",
}
`;

exports[`dbUtils findCoreEntity Find core entity without filters 1`] = `"FOR el IN core_trees RETURN el"`;

exports[`dbUtils findCoreEntity Find core entity without filters 2`] = `Object {}`;

exports[`dbUtils findCoreEntity Should filter label on any language 1`] = `"FOR el IN core_trees FILTER LIKE(el.@filterKey0, @filterValue0, true) FILTER el.@filterKey1 == @filterValue1 FILTER LIKE(el.label.fr, @filterValue2, true) OR LIKE(el.label.en, @filterValue2, true) RETURN el"`;

exports[`dbUtils findCoreEntity Should filter label on any language 2`] = `
Object {
  "filterKey0": "_key",
  "filterKey1": "system",
  "filterValue0": "categories",
  "filterValue1": false,
  "filterValue2": "Arbre des catégories",
}
`;
